### ISSUE TITLE
Show the %age translated on Textpacks

### ISSUE THREAD

**Issue by petecooper** [textpattern/textpattern#1604]
### Is your feature request related to a problem?

Nope.

### What is the feature?

Declare readiness of Textpacks and encourage language string translations by showing the percentage translated within the language bounding box, a bit like @opencollective do:

![Screenshot 2020-12-08 at 12 26 36](https://user-images.githubusercontent.com/226578/101484922-456ffa00-3952-11eb-9b54-34a7e4478173.png)

We don't hide the fact we like translators, and it's a neat way for users/admins to engage with a low barrier to entry, especially for multilingual native speakers.

**Comment by philwareham** [comment]
Just thinking how this could be done in a performant and bulletproof way.

Sure, you could simply count the lines in the en.ini file (if it exists in the `lang` directory) and compare against lines in the other languages - but I guess there is a remote possibility a user could add custom strings into their chosen language ini file. Also to do that we would have to read every one of these files to see how many lines are in each. That all seems clunky and intensive.

If this percentage score only happens once the strings are actually in the database, then it still looks like an intensive process to compare against the en.ini file.

@bloke @bloatware any thoughts?

**Comment by bloatware** [comment]
You could import them in a file from Crowdin?

**Comment by philwareham** [comment]
We could use the Crowdin API, but that tracks translation of files that we don't yet use in Textpattern (although might in future) such as the translated 'welcome to Textpattern' article (see https://github.com/textpattern/textpattern/issues/1179 for reference).

It also means a call to a third-party URL, so we'd have to have a fallback in place for if unavailable for various reasons).

**Comment by bloatware** [comment]
We'd do it only once, on each txp release and supply with?

**Comment by philwareham** [comment]
Oh, I see - yes that could work since we only officially update the languages with each release.

**Comment by Bloke** [comment]
Doing it in some offline capacity and stashing it somewhere on a release basis is fine by me. As long as people know it's not a real time value and is a snapshot from the last release, that's cool.

**Comment by bloatware** [comment]
We could update it when Diagnostic pings home?

**Comment by philwareham** [comment]
I could probably adjust the API stuff in https://textpattern.com/about/languages to allow that page to be used as an endpoint to check percentages. As long as we cache the API for a length of time (i.e. the API is not called every time someone pings the page). We have proposed a _roughly_ similar approach for the plugins site to check for updates to plugins.

**Comment by Bloke** [comment]
That would work. Fits in 4.9 with our improved handling/fallback of remote content (plugins, themes, etc). If we can guarantee to stay below their rate-limited connections, that seems a viable way forward.

**Comment by petecooper** [comment]
>We'd do it only once, on each txp release and supply with?

Yes – exactly this.

>We could update it when Diagnostic pings home?

No – unnecessary, unless I have misunderstood and we're thinking of dynamically updating language strings from within the admin-side, outside of a version update. Is that it?

**Comment by bloatware** [comment]
> No – unnecessary, unless I have misunderstood and we're thinking of dynamically updating language strings from within the admin-side, outside of a version update. Is that it?

Dunno. What is the purpose of %age - invitation to install or call for translation?

**Comment by petecooper** [comment]
>What is the purpose of %age - invitation to install or call for translation?

Both, but more the latter. I have a (possibly misguided) notion that if a Textpattern advocate sees their preferred language lacking, they may be inclined to assist with translation. A language pack with a relatively low percentage of translations will be cognitively more tricky from a UX standpoint (i.e. a mishmash of languages, see below), so a percentage shown might be an irrelevance anyway.

e.g. current state of ภาษาไทย (`th`):

![Screenshot 2021-02-14 at 18 19 29](https://user-images.githubusercontent.com/226578/107885303-58262280-6ef1-11eb-87bd-a3244c63f773.png)

**Comment by Bloke** [comment]
Any momentum on this?

If we have an API endpoint swinging near https://textpattern.com/about/languages that can return a nice JSON blob of current percentage translations, we can either run that just prior to release and hard-code it somewhere, or periodically check it every time someone visits the Languages panel - like we do with the check for updates on Diagnostics and the (proposed) check for updates to themes/plugins in 4.9.0.

**Comment by petecooper** [comment]
Should be easy enough to scrape the JSON (maybe a `jq` job) and build a SQL query that runs for each release, and as part of the upgrade scripts. I would perhaps avoid the on-demand check for dynamic percentages as that might imply an in-app upgrade route from our servers (or $translationservice servers, come to think on it), which is a whole other bundle of fun.

**Comment by Bloke** [comment]
True. If we're going fixed route, we could modify the header in each .ini file so it gets the percentage each time we push any changes:

```
[@common]
lang_name="English"
lang_code="en"
lang_dir="ltr"
lang_xlated="100"
[admin]
...
```

It's a bit of a hack though - and we'd have to ensure that entry doesn't get added to the language table as it's not a true string - so if anyone has any better ideas where to store this info (a single, delimited pref?) then throw them in the ring.

**Comment by petecooper** [comment]
Pondering this for 4.8.8… @Bloke, do you fancy a confab on this to see if it has legs?

Tagging https://github.com/textpattern/textpattern/issues/1720

**Comment by Bloke** [comment]
Definitely doable. Just gotta decide the best route wrt:

a) Opening an endpoint to get the values, preferably from content we control.
b) Storing the results in the distro somehow, given it's a one-shot-per-release event.

If we can fathom the most appropriate solutions for the above, it's a cinch to add to the languages panel, either in the dropdown of installed languages or, perhaps better(?), in each language cell so an admin knows the translation state prior to installation.

**Comment by Bloke** [comment]
I swear I put some stuff in place on the server to grab this value and expose it, but I can't remember if I did or if it's just on a local playground. Will try and locate the work.

**Comment by bloatware** [comment]
I have tried counting local files lines, it's quite fast. Sure, if a user alters some core language file, the %age will be biased, but I don't think many do, since custom strings would (?) be squashed on txp update.

**Comment by petecooper** [comment]
The %age translated pophelp could refer to CrowdIn to sidestep the chance of the local copy being ahead or behind (per the edge case above).

**Comment by Bloke** [comment]
Kind of related, but likely because we need to update the lang packs for 4.9.0 final. Or have we missed a string or two in en-gb and en-us?
![Screenshot 2024-09-10 at 00 33 25](https://github.com/user-attachments/assets/524ad828-043c-4fce-8e0d-091d43d15d6d)

**Comment by petecooper** [comment]
I picked up on the same thing earlier this evening. I diff'd the strings in `en` and `en-GB` and didn't find any missing…so I'm none the wiser.

**Comment by phiw13** [comment]
FWIW – 4.9-dev  with PHP 8.2.x, live server. But 4.9 beta1 shows 99% for `en-gb` and `en-us` , with PHP 8.3.x

<img width="610" alt="Screenshot 2024-09-10 at 11 18 51" src="https://github.com/user-attachments/assets/05ce0253-9e1d-4847-8f2e-13a8736eb60c">

**Comment by bloatware** [comment]
There seems to be few (`bulk_*` and `secondpass`) strings missing in `en-gb/us` beta pack.

**Comment by philwareham** [comment]
Hi @bloatware - please add any missing strings to [this issue](https://github.com/textpattern/textpacks/issues/313) - along with what group and a summary of what they mean. I'll then add to the correct language packs. Thanks!

**Comment by bloatware** [comment]
Hi @philwareham, I'm not sure it's the right place. These strings are already in the default `en` pack, it's more a language-specific issue.

**Comment by philwareham** [comment]
The English languages on the Textpack repo all have the same number of strings in them, and match what is in dev. Maybe I brought the latest packs in after the beta released.

**Comment by petecooper** [comment]
On the face of it, this looks like it's done in 4.9.0beta.2 - thanks to whoever made it happen!

Closing.

<img width="2628" height="800" alt="Image" src="https://github.com/user-attachments/assets/d068e1d9-25c4-4b9c-a19a-cc150d757b13" />
